Powpeg utxos spendability validation protocol (#419)
rsksmart/RSKIPs: IPs/RSKIP419.md

- Proposed Powpeg funding: the proposed Powpeg won't have funds to create the proof so the current Powpeg will create an UTXO covering the expenses.
- Proof transaction creation: once the proposed Powpeg receives the funds, it will create an UTXO to the current Powpeg to proof that it can spend them.
- Proof transaction verification: the current Powpeg will receive the proposed Powpeg spending-proof transaction, validate it was created by the proposed Powpeg from the funding transaction and proceed with the commitment.

#### Proposed Powpeg funding

Once the SVP starts, the Bridge will send the required satoshis to the proposed Powpeg address. It should be enough for it to be able to send back a transaction. Since the size of the Proof transaction will vary depending on the size of the proposed Powpeg input script, the calculation of the funds sent from the current Powpeg should consider that.

The process consists of the following 3 steps:

##### 1. Funding transaction creation

As soon as the proposed Powpeg has been voted (and saved in a new Bridge storage entry, key `proposedFederation`), the Bridge will create a SPV funding transaction for it.

> This transaction is similar to a peg-out where the recipient is the proposed Powpeg Bitcoin address, and the amount to receive is calculated based on the minimum amount of satoshis required to spend an UTXO by it.

Reviewer: This value should probably be calculated as `minPegoutValue + bitcoin fees`. So that we ensure the proposed powpeg signs a transaction worth `minPegoutValue`.

It can be tricky to calculate this though, if bitcoin fees are oscilating during the process.

Once the transaction is created, it has to be stored in the set for peg-outs waiting for confirmations, and the sighash has to be stored in the pegout index so the change could be registered.

Also, a new value in the storage should be added to identify this special kind of transaction:
    

>     KEY: "FUND_TX_HASH_UNSIGNED"

Reviewer: ```suggestion
    KEY: "SVP_FUND_TX_HASH_UNSIGNED"
```
What do you think? To give it context

Author: yes, i like it! thanks

    VALUE: hash of SVP funding tx without signatures

> 

Reviewer: Should we also mention here adding an output to a flyover federation built from  the proposed federation with a custom derivation hash?

Author: yes, absolutely. Thanks!

The SVP funding transaction goes through the usual peg-out confirmation process. Once the transaction accumulates at least 4_000 Rootstock blocks (for mainnet) it should be ready to be moved from the current set to the waiting for signatures set.

##### 3. Funding transaction signing

The funding transaction should be signed as any other peg-out. Once it is fully signed it should be removed from the set and a release_btc event should be emitted.
Eventually, when the funding transaction is registered in the Bridge, the Proof transaction could be created.

### Proof transaction creation

Once the funding transaction is registered, the Proof will be created for the proposed Powpeg to sign and broadcast it so it can prove it can spend funds.

##### Funding transaction registration
The funding transaction should be registered through the registerBtcTransaction method.
For that purpose, the method will be modified to be able to identify it.
It should verify:
- SVP period is ongoing
(i.e. proposedFederation exists & CURRENT_RSK_BLOCK < PROPOSED_FED_CREATION_BLOCK_NUMBER + SVP_PERIOD)
- Transaction hash without signatures matches the record in the FUND_TX_HASH_UNSIGNED storage entry

> Once everything is verified, the transaction will be registered and the transaction hash will be saved in a new storage entry, FUND_TX_HASH_SIGNED.

Reviewer: And remove `FUND_TX_HASH_UNSIGNED` value from storage? No longer needed and helps identify at which step of the process we currently are

The funding transaction hash signed being saved means that the proof Transaction can be created.

> **Important remark**: This method will be public, meaning any user could call it.

Reviewer: This got me a bit confused, sounds like we are talking about a new method. `registerBtcTransaction` already exists and is public

Author: you right, i'll just change it to `since this method is public,` instead of `this method will be public`
wdyt?

#### Proof transaction creation or validation process failure

A new method, `processPowpegChangeValidation` will be called from the `updateCollections` method.

The `processPowpegChangeValidation` will perform some actions if a proposedFederation exists depending on this two scenarios:

> - SVP period ended. This implies the validation was not successful.

Reviewer: ```suggestion
- SVP period ended and the proof transaction was not received. This implies the validation was not successful.
```

Author: I'm not sure about this one.
If the SVP period ended and the proposedFederation still exists, it implies that the proof tx was not received. Remember that after the proof tx is received, the proposedFederation won't exist any more.

wdyt?

Author: `SVP period ended. This implies the proof transaction was not received, and therefore that the validation was not successful.`

wdyt?

    -  The Bridge will emit a new event,
        

>         `COMMIT_FEDERATION_FAILED(blockNumber indexed uint)`

Reviewer: Do we need this value indexed?

Author: i dont know

Reviewer 2: good one, yes. Hope we won't use it but it could be useful for debugging if something happens

    - The proposed Powpeg will be eliminated.
    - The election will be allowed once again.

> - SVP period didn't end        

Reviewer: ```suggestion
- SVP period still ongoing
```

    - If the funding transaction signed hash is registered in the FUND_TX_HASH_SIGNED storage entry and the Proof transaction is not created yet, the Bridge will:
        - Create it as described [here](#proof-transaction)
        - Add it to a new storage set, key `spendTxWaitingForSignatures`
        - Set it in a new Storage entry:
            ```
            KEY: "SPEND_TX_HASH"

>             VALUE: hash of SVP proof tx

Reviewer: ```suggestion
            VALUE: hash of SVP proof tx without signatures
```

            ```    

##### Proof transaction

The Proof transaction is a Bitcoin transaction where an UTXO from the proposed Powpeg is sent to the current Powpeg.
This transaction will be as simple and small as possible to avoid spending much satoshis as fees.

Proof transaction structure:
```
{
    inputs: [
        {
            txid: FUND_TX_HASH_SIGNED,
            output_index: FUNDING_TX_OUTPUT_INDEX
        }
    ],
    outputs: [
        {
            script: CURRENT_POWPEG_OUTPUT_SCRIPT,
            value: TOTAL_UTXO - FEES
        }
    ]
}
```

#### Proof transaction confirmation

The SVP proof transaction should go through a new "peg-out signing process", since it has to be signed for the proposed pegnatories which belong neither the active nor the retiring powpeg, but the proposed one.

> #### Proof transaction signing

Reviewer: How is the powpeg gonna get the information of the proof transaction waiting for signatures? Should we update `getStateForRelease` method or it should it be a new method?

Author: i would say a new method is better

Author: but thats a hard part we have to discuss deeply

Reviewer 2: sounds good!

> The proof transaction should be signed as any other peg-out, but by the proposed powpeg. For this to happen, a new Bridge method `ADD_SPEND_TX_SIGNATURE` should be created to perform the exact same actions as the `ADD_SIGNATURE` but to accept signatures from public keys that don't belong to an active/retiring Powpeg (let's recall the proposed Powpeg is not active at this point).

Reviewer: ```suggestion
The proof transaction should be signed as any other peg-out, but by the proposed powpeg. For this to happen, a new Bridge method `addSpendTxSignature` should be created to perform the exact same actions as the `addSignature` but to accept signatures from public keys that don't belong to an active/retiring Powpeg (let's recall the proposed Powpeg is not active at this point).
```

The `addSpendTxSignature` method should check that the public key belongs to the proposed Powpeg and that the hash of the transaction intended to be signed exists in the `SPEND_TX_HASH` storage entry.

Once fully signed, the Bridge should remove the entry from the `spendTxWaitingForSignatures` and emit the release_btc event as usual.

### Proof verification

With the SVP proof transaction broadcasted, the Bridge now needs to validate it and proceed with the election commitment as expected.

> The proof transaction should be registered through the registerBtcTransaction method. As before, it should verify:

Reviewer: I'd prefer a new method, cleaner.

- SVP period is ongoing
- Transaction hash without signatures is registered in the `SPEND_TX_HASH` storage key.

> If the hash matches the one registered in the storage, then the SVP process is considered completed and we can proceed with the committment of the proposed -and now validated- Powpeg.

Reviewer: And remove SPEND_TX_HASH value from storage. Let's make sure we are cleaning up after ourselves as the process advances

Author: you right. and remove the SVP_FUND_TX_HASH_SIGNED from the storage right after the addition of the SVP_SPEND_TX_HASH

Eventually, when the funding transaction is registered in the Bridge, the Proof transaction could be created.

### Proof transaction creation

Once the funding transaction is registered, the Proof will be created for the proposed Powpeg to sign and broadcast it so it can prove it can spend funds.

##### Funding transaction registration
The funding transaction should be registered through the registerBtcTransaction method.
For that purpose, the method will be modified to be able to identify it.
It should verify:
- SVP period is ongoing
(i.e. proposedFederation exists & CURRENT_RSK_BLOCK < PROPOSED_FED_CREATION_BLOCK_NUMBER + SVP_PERIOD)
- Transaction hash without signatures matches the record in the SVP_FUND_TX_HASH_UNSIGNED storage entry

Once everything is verified, the transaction will be registered and the transaction hash will be saved in a new storage entry, SVP_FUND_TX_HASH_SIGNED. Also, the SVP_FUND_TX_HASH_UNSIGNED will be removed from the storage.
The funding transaction hash signed being saved means that the proof Transaction can be created.

> **Important remark**: this method is public, meaning any user could call it.

Reviewer 2: yes, I think it was more like an internal reminder. I'll delete it

Proof transaction structure:
```
{
    inputs: [
        {
            txid: SVP_FUND_TX_HASH_SIGNED,
            output_index: SVP_FUNDING_TX_OUTPUT_INDEX
        }
    ],
    outputs: [
        {
            script: CURRENT_POWPEG_OUTPUT_SCRIPT,
            value: TOTAL_UTXO - FEES
        }
    ]
}
```

#### Proof transaction confirmation

The SVP proof transaction should go through a new "peg-out signing process", since it has to be signed for the proposed pegnatories which belong neither the active nor the retiring powpeg, but the proposed one.

#### Proof transaction signing

The proof transaction should be signed as any other peg-out, but by the proposed powpeg. For this to happen, a new Bridge method `addSpendTxSignature` should be created to perform the exact same actions as the `addSignature` but to accept signatures from public keys that don't belong to an active/retiring Powpeg (let's recall the proposed Powpeg is not active at this point).
The `addSpendTxSignature` method should check that the public key belongs to the proposed Powpeg and that the hash of the transaction intended to be signed exists in the `SVP_SPEND_TX_HASH` storage entry.

> Once fully signed, the Bridge should remove the entry from the `spendTxWaitingForSignatures` and emit the release_btc event as usual.

Reviewer: Not sure about emitting `release_btc` event here. It is not a release or pegout, it is the proposed federation returning the funds. If someone is tracking that event it will create confusion

Reviewer 2: is true

### Proof verification

With the SVP proof transaction broadcasted, the Bridge now needs to validate it and proceed with the election commitment as expected.

The proof transaction should be registered through the registerBtcTransaction method. As before, it should verify:
- SVP period is ongoing
- Transaction hash without signatures is registered in the `SVP_SPEND_TX_HASH` storage key.

> If the hash matches the one registered in the storage, then the SVP process is considered completed and we can proceed with the committment of the proposed -and now validated- Powpeg.

Reviewer: And register the UTXO?

The `SVP_SPEND_TX_HASH` and the `proposedFederation` will be removed from storage.

### Validation phase duration

The duration of this phase must account for:
- Spending an UTXO from the current Powpeg (Bridge confirmations + pegnatories signing)
- Confirming the UTXO on Bitcoin (Bitcoin confirmations for the UTXO to be accepted)
- Sending back the UTXO to the current Powpeg (Proposed pegnatories signing)

The recommendation is that this phase takes approximately the blocks a peg-out confirmation has plus the bitcoin blocks needed to register the chance, twice. And a couple of thousands more blocks to be safe.
- Mainnet: 16_000
- Testnet: 80 - For testnet it is recommended to wait a bit longer as the Bitcoin miners don't tend to follow the 10' block creation time.
- Regtest: 15

### Bridge Storage changes

|Storage Key   |Type          |Description   |
|:------------ |:-------------|:-------------|
|SVP_FUND_TX_HASH_UNSIGNED|bytes32|hash of SVP funding tx unsigned|
|SVP_FUND_TX_HASH_SIGNED|bytes32|hash of SVP funding tx signed|

> |SVP_SPEND_TX_HASH|bytes32|hash of SVP proof tx|

Reviewer: Change these keys to camelCase

Reviewer: Would look better putting the signature
```
svpSpendTxHash(bytes32 indexed btcTxHash)
```

### New Bridge methods
|Method   |Type          |Description   |
|:------------ |:-------------|:-------------|

> |ADD_SPEND_TX_SIGNATURE|void|method to sign the proof transaction by the proposed pegnatories|

Reviewer: Same here
